base/tuple.jl

```diff
@@ -29,29 +29,43 @@ eltype{T,_}(::Type{NTuple{_,T}}) = T
 
 ## mapping ##
 
-ntuple(f::Function, n::Integer) =
+const ntuple = function(f, n::Integer)
     n<=0 ? () :
     n==1 ? (f(1),) :
     n==2 ? (f(1),f(2),) :
     n==3 ? (f(1),f(2),f(3),) :
-    n==4 ? (f(1),f(2),f(3),f(4),) :
-    n==5 ? (f(1),f(2),f(3),f(4),f(5),) :
-    tuple(ntuple(f,n-5)..., f(n-4), f(n-3), f(n-2), f(n-1), f(n))
-
-ntuple(f, ::Type{Val{0}}) = ()
-ntuple(f, ::Type{Val{1}}) = (f(1),)
-ntuple(f, ::Type{Val{2}}) = (f(1),f(2))
-ntuple(f, ::Type{Val{3}}) = (f(1),f(2),f(3))
-ntuple(f, ::Type{Val{4}}) = (f(1),f(2),f(3),f(4))
-ntuple(f, ::Type{Val{5}}) = (f(1),f(2),f(3),f(4),f(5))
-@generated function ntuple{N}(f, ::Type{Val{N}})
-    if !isa(N,Int)
-        :(throw(TypeError(:ntuple, "", Int, $(QuoteNode(N)))))
+    tuple([f(i) for i = 1:Int(n)]...)
```

Comment:
i haven't looked into the exact issues in #13359 to see which point in codegen is pessimizing its assumptions for that example. but this is calling `tuple`, which is a more like syntax than an actual function call. the tfunc below then tries to do a better job of computing the type information that will be needed by the callee to limit the loss of type information in the caller where it could cause a much greater slowdown.